core/drivers/clk/clk_dt.c

```diff
@@ -0,0 +1,368 @@
+
+		clk_drv = clk_get_compatible_driver(compat);
+		if (clk_drv)
+			clk_setup_compatible(fdt, compat, clk_drv);
+	}
+}
+
+static void parse_clock_property(const void *fdt, int nodeoffset)
+{
+	int len = 0;
+	int idx = 0;
+	int parent_node = 0;
+	int clock_cells = 0;
+	uint32_t phandle = 0;
+	const uint32_t *prop = NULL;
+
+	prop = fdt_getprop(fdt, nodeoffset, "clocks", &len);
+	if (!prop)
+		return;
+
+	while (idx < len) {
+		phandle = fdt32_to_cpu(prop[idx]);
+
+		parent_node = fdt_node_offset_by_phandle(fdt, phandle);
+		if (parent_node < 0)
+			return;
+
+		probe_parent_clock(fdt, parent_node);
+
+		clock_cells = fdt_clock_cells(fdt, parent_node);
+		idx += sizeof(phandle) + clock_cells * sizeof(uint32_t);
+	}
+}
+
+static void clk_setup_compatible(const void *fdt, const char *compatible,
+				 const struct clk_driver *clk_drv)
+{
+	int status = 0;
+	TEE_Result res;
+	int node = fdt_node_offset_by_compatible(fdt, -1, compatible);
+
+	if (node < 0 || clk_get_provider(node))
+		return;
+
+	while (node >= 0) {
+		status = _fdt_get_status(fdt, node);
+		if (!(status & DT_STATUS_OK_SEC))
+			continue;
+
+		parse_clock_property(fdt, node);
+
+		res = clk_drv->setup(fdt, node);
+		if (res != TEE_SUCCESS) {
+			EMSG("Failed to setup clock driver for compatible %s",
+			      compatible);
+			panic();
+		}
+
+		node = fdt_node_offset_by_compatible(fdt, node, compatible);
+	};
+}
+
+static bool clk_dt_set_parent(struct clk *clk, struct clk *parent)
+{
+	int i = 0;
+
+	for (i = 0; i < clk_get_num_parents(clk); i++) {
+		if (clk_get_parent_by_index(clk, i) == parent) {
+			clk_set_parent(clk, i);
+			return true;
+		}
+	}
+
+	EMSG("Clock %s is not a parent of clock %s", parent->name, clk->name);
+	return false;
+}
+
+static void parse_assigned_clock(const void *fdt, int nodeoffset)
+{
+	int rate_len = 0;
+	int clock_idx = 0;
+	struct clk *clk = NULL;
+	unsigned long rate = 0;
+	struct clk *parent = NULL;
+	const uint32_t *rate_prop = NULL;
+
+	rate_prop = fdt_getprop(fdt, nodeoffset, "assigned-clock-rates",
+				&rate_len);
+	if (rate_len)
+		rate_len /= sizeof(uint32_t);
+
+	while (1) {
+		clk = clk_dt_get_by_idx_prop("assigned-clocks", fdt, nodeoffset,
+					     clock_idx);
+		if (!clk)
+			return;
+
+		parent = clk_dt_get_by_idx_prop("assigned-clock-parents", fdt,
+						nodeoffset, clock_idx);
+
+		if (parent) {
+			if (!clk_dt_set_parent(clk, parent))
+				return;
+		}
+
+		if (rate_prop && clock_idx <= rate_len) {
+			rate = fdt32_to_cpu(rate_prop[clock_idx]);
+			if (rate)
+				clk_set_rate(clk, rate);
+		}
+
+		clock_idx++;
+	}
+}
+
+static void clk_probe_assigned(const void *fdt, int parent_node)
+{
+	int len = 0;
+	int child = 0;
+	int status = 0;
+
+	fdt_for_each_subnode(child, fdt, parent_node) {
+		clk_probe_assigned(fdt, child);
+
+		status = _fdt_get_status(fdt, child);
+		if (!status)
+			continue;
+
+		if (fdt_getprop(fdt, child, "assigned-clocks", &len))
+			parse_assigned_clock(fdt, child);
+	}
+}
+
+TEE_Result clk_dt_probe(void)
```

Comment:
I think you can discard commit "kernel: boot: probe clocks before console" and call `clk_dt_probe()` from `early_init()` initcall level. There is no need to call it from `boot_init_primary_late()` specifically, do we? Note the console config is done from `boot_init_primary_late()` because it initially only accessed the 'external dt' and every external_dt accessed where called from `boot_init_primary_late()`.

If console driver needs clocks then it's an issue. What about a console driver that would need clock and pin control. and maybe  a reset controller driver. I think it would not be nice to have all these specific drivers initialized from `boot_init_primary_late()` right before `configure_console_from_dt()` is called.
What do you think?

Reply:
Maybe the console are no exception and should be probed later also ? For instance, having an early console is needed to allow debugging but once done, OP-TEE could probably use a more full featured/secure one using devicetree.
But you are right, console probably need (at least) to set pinctrl and clocks to correctly work.

So in the meantime until this is sorted out, I'll reswitch to early_init ;)

Reply:
> Maybe the consoles are no exception and should be probed later also ? For instance, having an early console is needed to allow debugging 

True, that is hard-coded per-platform with various degrees of configure-ability.

> but once done, 
> OP-TEE could probably use a more full featured/secure one using devicetree.

We have it more or less, see [here](https://github.com/OP-TEE/optee_os/commit/4dc31c52544a7241e97d0452d512330ee30ed664) and [here](https://github.com/OP-TEE/optee_os/commit/08baa8c9ae08f355dfd9934a86e8d9dbc913c5a1)). However...

> But you are right, console probably need (at least) to set pinctrl and clocks to correctly work.

...not sure how we would achieve this.